src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala

```diff
@@ -1221,8 +1221,17 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
         case Some(jar: ZipArchive) => // We are in a jar
```

Comment:
the scala 2 compiler uses generally the same io/classpath code as scala 3 (Scala 3 copied it from Scala 2). The classpath for Zip entries is cached and invalidated based on the underlying filesystem change, so I suggest instead of creating a classloader and loading resource, this whole method should instead do something similar to 
`file.container.lookupName(file.name.stripSuffix(".class") + ".tasty", directory = false)` and then check for null

Reply:
alternatively, only do that lookup in this branch

Reply:
I did some digging and I'm not totally confident using the `AbstractFile` methods will work well here for the `ZipArchive` branch. This is because most of the methods like `container`, `lookupName`, etc. are `unsupported` in `ZipArchive` https://github.com/scala/scala/blob/2.13.x/src/reflect/scala/reflect/io/ZipArchive.scala#L107

If we wanted to just support `FileZipArchive` we could use `root` from that class https://github.com/scala/scala/blob/2.13.x/src/reflect/scala/reflect/io/ZipArchive.scala#L248

But `root` is a `lazy val` and it requires walking the entire jar file to create a `DirEntry` that represents the jar, so I worry a bit about performance there if all we want is a single file out of it.

`DirEntry` does have a `lookupName` function: https://github.com/scala/scala/blob/2.13.x/src/reflect/scala/reflect/io/ZipArchive.scala#L130 That being said, as far as I can tell that function it only goes one directory at a time. So we'd need to split the tasty file path into its components and call `lookupName` for all the directories on the way to the tasty file, and then one final time for the tasty file itself.

This is all doable, but I wanted to run it by by you before actually writing the code. Part of me thinks that just fixing the bug in the existing `URLClassLoader` approach may make more sense here. However, I will totally defer to you as you're the expert here and my analysis above may be totally off.

Reply:
Here is what we used to do in dotty, is this portable? https://github.com/lampepfl/dotty/blame/c62909009506171b03e092c4a391483b044d3b49/compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala#L951-L974

Reply:
Sorry for the delay, I got caught up in a few other things going on at work. I have not forgotten about this PR, though and it is still a priority for me. I hope to have updates in the next few days.

Reply:
ok, thank you for the update :)